### DISCUSSION TITLE
working with and developing glazedlists. Community?

### DISCUSSION THREAD

**Discussion by errael** [glazedlists/glazedlists#698] [Q&A]
The project seems abandoned. (I'm going to look through the forks, but ...)

I'm going to work on a fix for #676. I'm wondering if the project might be active in some other location.

I'm basing my fix on 1.11.0. I was tempted to back up to 1.10.0 since I have no gradle experience. Could go with 1.12 snap, but I wonder about stability; it looks mostly like jdk1.8+ support.

Any insights appreciated.

**Comment by hbrands** [comment]
Hi,
this is the main project for Glazed Lists.
There are a few maintainers who work on Glazed Lists in their free time.
There has not been much activity in the last months, but it's not abandoned.
We intend to release version 1.12 "soonish", which is currently on the master branch.
It's quite stable, I'm using it in a very big production application for example.
So I suggest to base your work on master. Thanks!
Holger

**Comment by errael** [comment]
I'm using Netbeans. I can open/edit/debug core, but I can't run individual tests from NetBeans. For example I copied AutoCompleteSupportTestApp to a separate NetBeans project and can run/debug that way. If I open other subprojects, for example extensions/icu4j, when I open a file, GlazedListsICU4J.java for example, the imports that resolve into the "core" jar are not found (error highlights from IDE). I'm investigating this from the NetBeans side.

I suspect the rather archaic build.gradle files are at the root of the issue. I want to add some tests for AutoCompleteSupport. The tests of interest are in extensions/icu4j, for example AutoCompleteSupportTest. I'm not doing anything in particular with icu4j, but that's where the test is.

Since this is the first project I've worked with that uses gradle, I'm at more of a loss than usual. Any hints or suggestions appreciated.

**Comment by errael** [comment]
"./gradlew build --scan" produced https://gradle.com/s/iwj7ksmm72wc2 with deprecation
```
The configuration :extensions:compileClasspath was resolved without accessing the project in a safe manner.
```
This seems related to the type of problem I'm having, so I'll see what I can see. The deprecation warning suggested looking at
```
https://docs.gradle.org/5.3.1/userguide/troubleshooting_dependency_resolution.html#sub:configuration_resolution_constraints
```

**Comment by hbrands** [comment]
> I'm using Netbeans. I can open/edit/debug core, but I can't run individual tests from NetBeans. For example I copied AutoCompleteSupportTestApp to a separate NetBeans project and can run/debug that way. If I open other subprojects, for example extensions/icu4j, when I open a file, GlazedListsICU4J.java for example, the imports that resolve into the "core" jar are not found (error highlights from IDE). I'm investigating this from the NetBeans side.
> 
> I suspect the rather archaic build.gradle files are at the root of the issue. I want to add some tests for AutoCompleteSupport. The tests of interest are in extensions/icu4j, for example AutoCompleteSupportTest. I'm not doing anything in particular with icu4j, but that's where the test is.
> 
> Since this is the first project I've worked with that uses gradle, I'm at more of a loss than usual. Any hints or suggestions appreciated.

Sorry, I'm not familiar with Netbeans, I'm a Eclipse user. Usually there are IDE plugins for supporting build tools, like m2e for Maven support and Buildship for Gradle support in Eclipse. 
I assume there is some kind of plugin for Gradle support in Netbeans, too. A quick Google search revealed this link:
https://netbeans.org/features/java/build-tools.html
Hope this helps.

**Comment by hbrands** [comment]
@errael FYI: I've enabled the Discussions feature of Github for this project and moved your question to the Q&A section.

**Comment by errael** [comment]
I'm using NetBeans support for gradle, but I'm guessing it's more restrictive. I'm trying to figure out how to get rid of the warning about the extensions classpath resolution. I'm hoping that fixing that will fix the NetBeans issue.

Fortunately the tests I'm working on can be put together with minimal scaffolding, `assertEquals()` and some catching; then move them into their proper place.

**Comment by errael** [comment]
The issue with NetBeans is solved.

I run NB on jdk-15, so that's the default to run gradle. Gradle 5.3.1 (used by glazedlists) doesn't support newer jdks and that was the root of the problems I was having. Set some properties in NetBeans so that it uses jdk-11 to run gradle; no more error badges in the editor and I can use the IDE to edit and run tests, even "debug focused test method" is working.

**Reply by hbrands** [reply]
I'm glad you could resolve this issue.

Please keep in mind that Glazed Lists 1.12 will have a Java 8 baseline.
(If you try to develop with a JDK 8, be sure to use one with JavaFx bundled, e.g. an Oracle Java 8 JDK)

**Reply by errael** [reply]
> Glazed Lists 1.12 will have a Java 8 baseline

(gradle novice here) I've assumed that `javaTargetVersion = 1.8` does that.

**Reply by errael** [reply]
> If you try to develop with a JDK 8, be sure to use one with JavaFx bundled, e.g. an Oracle Java 8 JDK

I don't understand the implications. 

Does "develop" mean use? Is this about producing a compatible glazedlist artifact?
Does this mean that glazedlists won't work in a project running on jdk11 or later? 
I'm guessing this means using `javac --release 8 ...` from a jdk-9+ doesn't work.
Is this only if you want to use javafx? Is there a way to get javafx to work on a recent java?
Does the future glazedlists-2.0 address this?

**Reply by hbrands** [reply]
javaTargetVersion = 1.8 ensures that source and target is set to 1.8, indeed. But when you build against a JDK 11 (as you indicated above) it would still be possible to use a Java 11 API unnoticed. 
(Of course, I would notice because I'll build the pull request against Java 8, so no big problem.)
You are right, using the -release flag would not have these problems, but this release flag is only supported in Gradle >= 6.6 I think.
(A major upgrade of Gradle is planned for GL 2.0)

Concerning JavaFx: JavaFx is only needed for the Glazed Lists JavaFx extension. So, building all parts of Glazed Lists, the JavaFx APIs are required on the classpath. Oracle Java 8 JDK bundles JavaFx, so that should work out of the box. Building against a JDK 8 without bundled JavaFx (e.g. from AdoptOpenJDK) will not work out of the box.

Things changed with newer JDK versions. Oracle has removed JavaFx from its JDK, so JavaFx is usually included as any other third-party dependency. Therefore, we use the org.openjfx.javafxplugin Gradle plugin for Java >= 11 to include JavaFx as dependency. Please see the build.gradle file of the JavaFx extension for details.

So, to sum up: if you want to use and build against a Java 8 version, please use a Oracle JDK 8.
Using Java versions >= 11, you are not forced to use a Oracle JDK, use your favorite JDK distribution.
(But make sure to use only Java APIs <= 1.8 for Glazed Lists 1.12.)
Hope this helps.

**Reply by errael** [reply]
> Hope this helps.

Very much so. Thanks.

BTW, I do a command line build, with JAVA_HOME jdk1.8 oracle, as a final check. (Of course sh*t happens)
